universal-application-tool-0.0.1/app/views/admin/questions/QuestionEditView.java

```diff
@@ -0,0 +1,117 @@
+package views.admin.questions;
+
+import static j2html.TagCreator.body;
+import static j2html.TagCreator.br;
+import static j2html.TagCreator.form;
+import static j2html.TagCreator.label;
+
+import com.google.common.collect.ImmutableList;
+import com.google.inject.Inject;
+import j2html.tags.ContainerTag;
+import j2html.tags.DomContent;
+import java.util.AbstractMap.SimpleEntry;
+import java.util.Arrays;
+import java.util.Locale;
+import java.util.Optional;
+import play.mvc.Http.Request;
+import play.twirl.api.Content;
+import services.question.QuestionDefinition;
+import services.question.QuestionType;
+import services.question.TranslationNotFoundException;
+import views.BaseHtmlLayout;
+import views.BaseHtmlView;
+
+public final class QuestionEditView extends BaseHtmlView {
+  private final BaseHtmlLayout layout;
+
+  @Inject
+  public QuestionEditView(BaseHtmlLayout layout) {
+    this.layout = layout;
+  }
+
+  public Content render(Request request, Optional<QuestionDefinition> question) {
+    String headerText = question.isPresent() ? "Edit Question" : "New Question";
+    return layout.htmlContent(
+        body(
+            renderHeader(headerText),
+            buildQuestionForm(question).with(makeCsrfTokenInputTag(request))));
+  }
+
+  private ContainerTag buildQuestionForm(Optional<QuestionDefinition> optionalDefinition) {
```

Comment:
suggestion (non-blocking): Would this be clearer as two methods, instead of trying to combine the `create new` and `edit` form into one method?

Reply:
I feel like it'd be a lot of repeated code. The biggest difference is just whether we have data to populate the inputs. I think that we could probably reuse `QuestionForm` to clean it up a bit, though.

Reply:
suggestion (blocking): you could factor out the part that adds `QuestionForm` fields to the `formTag` and split this into an `edit` and `create` method. 

I find the optional input confusing. It would be much clearer to me if there were a `render()` and a `render(QuestionDefinition)`, and `buildquestionForm()` and `buildQuestionForm(QuestionDefinition)`.